src/app/bdx/DiagnosticLogsBDXTransferHandler.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/core/CHIPConfig.h>
+
+#if CHIP_CONFIG_ENABLE_BDX_LOG_TRANSFER
+#include "DiagnosticLogsBDXTransferHandler.h"
+#include <app/clusters/diagnostic-logs-server/diagnostic-logs-server.h>
+#include <messaging/ExchangeMgr.h>
+#include <platform/CHIPDeviceLayer.h>
+
+using chip::BitFlags;
+using chip::ByteSpan;
+using chip::CharSpan;
+using chip::FabricIndex;
+using chip::FabricInfo;
+using chip::MutableCharSpan;
+using chip::NodeId;
+using chip::Span;
+using chip::bdx::TransferControlFlags;
+using chip::Protocols::InteractionModel::Status;
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::bdx;
+using namespace chip::app::Clusters::DiagnosticLogs;
+
+// BDX Transfer Params
+constexpr System::Clock::Timeout kBdxPollIntervalMs = System::Clock::Milliseconds32(50);
+
+// Timeout for the BDX transfer session. The OTA Spec mandates this should be >= 5 minutes.
+constexpr System::Clock::Timeout kBdxTimeout = System::Clock::Seconds16(5 * 60);
+
+constexpr uint16_t kBdxMaxBlockSize = 1024;
+
+CHIP_ERROR DiagnosticLogsBDXTransferHandler::InitializeTransfer(chip::Messaging::ExchangeContext * exchangeCtx,
+                                                                FabricIndex fabricIndex, NodeId nodeId,
+                                                                LogProviderDelegate * delegate, IntentEnum intent,
+                                                                CharSpan fileDesignator)
+{
+    if (mInitialized)
+    {
+        // Reset stale connection from the same Node if exists.
+        Reset();
+    }
+
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    mExchangeCtx = exchangeCtx->GetExchangeMgr()->NewContext(exchangeCtx->GetSessionHandle(), this);
```

Comment:
i think I originally had the exchange manager and session handle and then tried to optimize lets pass one object instead of 2. :) Makes sense to pass it separately and make it clearer. will add some comments as well.

Reply:
done.